File: PRE/grocery-bot-python/pathfinding.py
"""Pathfinding - optimized with flat arrays and tuples."""
from __future__ import annotations
from collections import deque

from types_ import (
    FLOOR, WALL, SHELF, DROPOFF,
    UP, DOWN, LEFT, RIGHT,
    DIR_DX, DIR_DY,
    UNREACHABLE,
)

# Reusable direction offsets
_DDX = (0, 0, -1, 1)
_DDY = (-1, 1, 0, 0)
_DIRS = (UP, DOWN, LEFT, RIGHT)
_ADJ_OFFSETS = ((0, -1), (0, 1), (-1, 0), (1, 0))


def bfs_dist_map(grid, w, h, sx, sy):
    """BFS distance map from (sx,sy). Returns flat list[int] of size w*h."""
    size = w * h
    dm = [UNREACHABLE] * size

    if sx < 0 or sy < 0 or sx >= w or sy >= h:
        return dm
    idx = sy * w + sx
    cell = grid[idx]
    if cell == WALL or cell == SHELF:
        return dm

    dm[idx] = 0
    queue = deque()
    queue.append(idx)

    while queue:
        ci = queue.popleft()
        cy, cx = divmod(ci, w)
        # Optimized: inline divmod result
        cy = ci // w
        cx = ci - cy * w
        cd = dm[ci]
        nd = cd + 1
        for i in range(4):
            nx = cx + _DDX[i]
            ny = cy + _DDY[i]
            if nx < 0 or ny < 0 or nx >= w or ny >= h:
                continue
            ni = ny * w + nx
            if dm[ni] != UNREACHABLE:
                continue
            c = grid[ni]
            if c == WALL or c == SHELF:
                continue
            dm[ni] = nd
            queue.append(ni)

    return dm


def bfs_path(grid, w, h, sx, sy, tx, ty, bot_id, bot_positions, bot_count):
    """BFS with first-step collision avoidance. Returns (dist, first_dir or -1)."""
    if sx == tx and sy == ty:
        return (0, -1)

    size = w * h
    # Use bytearray as visited (faster than set for small grids)
    visited = bytearray(size)
    si = sy * w + sx
    visited[si] = 1

    # Queue: flat list of (index, dist, first_dir)
    queue = deque()
    queue.append((si, 0, -1))

    while queue:
        ci, cd, cfd = queue.popleft()
        cy = ci // w
        cx = ci - cy * w
        for i in range(4):
            nx = cx + _DDX[i]
            ny = cy + _DDY[i]
            if nx < 0 or ny < 0 or nx >= w or ny >= h:
                continue
            ni = ny * w + nx
            if visited[ni]:
                continue

            cell = grid[ni]
            is_target = (nx == tx and ny == ty)
            if not is_target and (cell == WALL or cell == SHELF):
                continue

            # First-step collision avoidance
            if cd == 0:
                blocked = False
                for bi in range(bot_count):
                    if bi == bot_id:
                        continue
                    bx, by = bot_positions[bi]
                    if bx == nx and by == ny:
                        blocked = True
                        break
                if blocked:
                    continue

            visited[ni] = 1
            first = cfd if cfd >= 0 else _DIRS[i]
            if is_target:
                return (cd + 1, first)
            queue.append((ni, cd + 1, first))

    # Fallback: greedy direction
    return (UNREACHABLE, _safe_greedy(grid, w, h, sx, sy, tx, ty))


def _safe_greedy(grid, w, h, fx, fy, tx, ty):
    """Greedy direction toward target."""
    ddx = tx - fx
    ddy = ty - fy

    if abs(ddx) > abs(ddy):
        ordered = (RIGHT, DOWN, UP, LEFT) if ddx > 0 else (LEFT, DOWN, UP, RIGHT)
    elif ddy > 0:
        ordered = (DOWN, RIGHT, LEFT, UP)
    elif ddy < 0:
        ordered = (UP, RIGHT, LEFT, DOWN)
    else:
        ordered = (RIGHT, DOWN, LEFT, UP)

    for d in ordered:
        nx = fx + DIR_DX[d]
        ny = fy + DIR_DY[d]
        if nx < 0 or ny < 0 or nx >= w or ny >= h:
            continue
        cell = grid[ny * w + nx]
        if cell == FLOOR or cell == DROPOFF:
            return d
    return -1


def find_best_adj(grid, w, h, ix, iy, dm):
    """Find best adjacent floor cell to item at (ix,iy) using distance map dm.
    Returns (ax, ay) or None."""
    best = None
    best_d = UNREACHABLE

    for ox, oy in _ADJ_OFFSETS:
        nx, ny = ix + ox, iy + oy
        if nx < 0 or ny < 0 or nx >= w or ny >= h:
            continue
        cell = grid[ny * w + nx]
        if cell == FLOOR or cell == DROPOFF:
            d = dm[ny * w + nx]
            if d < best_d:
                best_d = d
                best = (nx, ny)

    return best

File: PRE/grocery-bot-zig/src/pathfinding.zig
const std = @import("std");
const types = @import("types.zig");

const Pos = types.Pos;
const Dir = types.Dir;
const Cell = types.Cell;
const GameState = types.GameState;
const DistMap = types.DistMap;
const MAX_H = types.MAX_H;
const MAX_W = types.MAX_W;
const MAX_CELLS = types.MAX_CELLS;
const MAX_BOTS = types.MAX_BOTS;
const UNREACHABLE = types.UNREACHABLE;

// ── BFS Distance Map ──────────────────────────────────────────────────
pub fn bfsDistMap(state: *const GameState, start: Pos, dm: *DistMap) void {
    const w = state.width;
    const h = state.height;
    for (0..h) |y| for (0..w) |x| {
        dm[y][x] = UNREACHABLE;
    };
    if (start.x < 0 or start.y < 0 or start.x >= w or start.y >= h) return;
    const cell_s = state.grid[@intCast(start.y)][@intCast(start.x)];
    if (cell_s == .wall or cell_s == .shelf) return;

    var queue: [MAX_CELLS]struct { x: u16, y: u16 } = undefined;
    var head: u16 = 0;
    var tail: u16 = 0;
    const sx: u16 = @intCast(start.x);
    const sy: u16 = @intCast(start.y);
    dm[sy][sx] = 0;
    queue[tail] = .{ .x = sx, .y = sy };
    tail += 1;

    const ddx = [4]i16{ 0, 0, -1, 1 };
    const ddy = [4]i16{ -1, 1, 0, 0 };
    while (head < tail) {
        const cur = queue[head];
        head += 1;
        const cd = dm[cur.y][cur.x];
        for (0..4) |i| {
            const nx_i = @as(i32, cur.x) + ddx[i];
            const ny_i = @as(i32, cur.y) + ddy[i];
            if (nx_i < 0 or ny_i < 0 or nx_i >= w or ny_i >= h) continue;
            const nx: u16 = @intCast(nx_i);
            const ny: u16 = @intCast(ny_i);
            if (dm[ny][nx] != UNREACHABLE) continue;
            const cell = state.grid[ny][nx];
            if (cell == .wall or cell == .shelf) continue;
            dm[ny][nx] = cd + 1;
            queue[tail] = .{ .x = nx, .y = ny };
            tail += 1;
        }
    }
}

// ── Distance lookup from a precomputed DistMap ────────────────────────
pub fn distFromMap(dm: *const DistMap, pos: Pos) u16 {
    if (pos.x < 0 or pos.y < 0) return UNREACHABLE;
    return dm[@intCast(pos.y)][@intCast(pos.x)];
}

// ── BFS with first-step collision avoidance ────────────────────────────
pub const BfsResult = struct { dist: u16, first_dir: ?Dir };

pub fn bfs(state: *const GameState, start: Pos, target: Pos, bot_id: u8, bot_positions: *const [MAX_BOTS]Pos) BfsResult {
    if (start.eql(target)) return .{ .dist = 0, .first_dir = null };
    const w = state.width;
    const h = state.height;
    var visited: [MAX_CELLS]bool = undefined;
    @memset(visited[0 .. w * h], false);
    const QEntry = struct { x: u16, y: u16, dist: u16, first_dir: ?Dir };
    var queue: [MAX_CELLS]QEntry = undefined;
    var head: u16 = 0;
    var tail: u16 = 0;
    const sx: u16 = @intCast(start.x);
    const sy: u16 = @intCast(start.y);
    visited[sy * w + sx] = true;
    queue[tail] = .{ .x = sx, .y = sy, .dist = 0, .first_dir = null };
    tail += 1;

    const ddx = [4]i16{ 0, 0, -1, 1 };
    const ddy = [4]i16{ -1, 1, 0, 0 };
    const dirs = [4]Dir{ .up, .down, .left, .right };
    while (head < tail) {
        const cur = queue[head];
        head += 1;
        for (0..4) |i| {
            const nx_i = @as(i32, cur.x) + ddx[i];
            const ny_i = @as(i32, cur.y) + ddy[i];
            if (nx_i < 0 or ny_i < 0 or nx_i >= w or ny_i >= h) continue;
            const nx: u16 = @intCast(nx_i);
            const ny: u16 = @intCast(ny_i);
            const idx = ny * w + nx;
            if (visited[idx]) continue;
            const cell = state.grid[ny][nx];
            const is_target = (nx == @as(u16, @intCast(target.x)) and ny == @as(u16, @intCast(target.y)));
            if (!is_target and (cell == .wall or cell == .shelf)) continue;
            if (cur.dist < 2) {
                var blocked = false;
                for (0..state.bot_count) |bi| {
                    if (bi == bot_id) continue;
                    if (bot_positions[bi].x == @as(i16, @intCast(nx)) and bot_positions[bi].y == @as(i16, @intCast(ny))) {
                        blocked = true;
                        break;
                    }
                }
                if (blocked) continue;
            }
            visited[idx] = true;
            const first = if (cur.first_dir) |d| d else dirs[i];
            if (is_target) return .{ .dist = cur.dist + 1, .first_dir = first };
            queue[tail] = .{ .x = nx, .y = ny, .dist = cur.dist + 1, .first_dir = first };
            tail += 1;
        }
    }
    // BFS failed - try to find any walkable direction as fallback
    return .{ .dist = UNREACHABLE, .first_dir = safeGreedyDir(state, start, target) };
}

pub fn safeGreedyDir(state: *const GameState, from: Pos, to: Pos) ?Dir {
    const ddx2 = to.x - from.x;
    const ddy2 = to.y - from.y;
    const ordered_dirs: [4]Dir = blk: {
        if (@abs(ddx2) > @abs(ddy2)) {
            if (ddx2 > 0) break :blk [4]Dir{ .right, .down, .up, .left };
            break :blk [4]Dir{ .left, .down, .up, .right };
        }
        if (ddy2 > 0) break :blk [4]Dir{ .down, .right, .left, .up };
        if (ddy2 < 0) break :blk [4]Dir{ .up, .right, .left, .down };
        break :blk [4]Dir{ .right, .down, .left, .up };
    };
    const offx = [4]i16{ 0, 0, -1, 1 };
    const offy = [4]i16{ -1, 1, 0, 0 };
    for (ordered_dirs) |d| {
        const di: usize = switch (d) { .up => 0, .down => 1, .left => 2, .right => 3 };
        const nx = from.x + offx[di];
        const ny = from.y + offy[di];
        if (nx < 0 or ny < 0 or nx >= state.width or ny >= state.height) continue;
        const cell = state.grid[@intCast(ny)][@intCast(nx)];
        if (cell == .floor or cell == .dropoff) return d;
    }
    return null;
}

// ── Find best adjacent floor cell to an item ──────────────────────────
pub fn findBestAdj(state: *const GameState, item_pos: Pos, dm: *const DistMap) ?Pos {
    const offsets = [4][2]i16{ .{ 0, -1 }, .{ 0, 1 }, .{ -1, 0 }, .{ 1, 0 } };
    var best: ?Pos = null;
    var best_d: u16 = UNREACHABLE;
    for (offsets) |off| {
        const nx = item_pos.x + off[0];
        const ny = item_pos.y + off[1];
        if (nx < 0 or ny < 0 or nx >= state.width or ny >= state.height) continue;
        const cell = state.grid[@intCast(ny)][@intCast(nx)];
        if (cell == .floor or cell == .dropoff) {
            const d = dm[@intCast(ny)][@intCast(nx)];
            if (d < best_d) { best_d = d; best = .{ .x = nx, .y = ny }; }
        }
    }
    return best;
}
